### ISSUE TITLE
wss gets 403 on console

### ISSUE THREAD

**Issue by Mangirdaz** [openshift/origin#12222]
We get ws 403 errors on console with user token

##### Version
Server https://console.int:443
openshift v3.3.0.35
kubernetes v1.3.0+52492b4

##### Steps To Reproduce
1. We run AVI load balancer as our LB for Openshift. 
2. Create VIP with static DNS record and pool members are set Masters
3. Session affinity is set to client IP

p.s Token is valid. 
##### Current Result
![image](https://cloud.githubusercontent.com/assets/3225744/21102162/db87912a-c074-11e6-9480-70c48f90a1b1.png)

##### Expected Result
Session is kept alive.

I dont expect solution, but maybe you seen this somewhere else or know how to debug it...

**Comment by jwforres** [comment]
Is it possible something in your load balancer is stripping query params off the request?

@liggitt any thoughts on this?

**Comment by jwforres** [comment]
or @deads2k since @liggitt is OOO

**Comment by Mangirdaz** [comment]
We do SSL termination on AVI VIP. So we still suspect it may be issue on How AVI Load Balancer terminates SSL...  Still bumpt

**Comment by deads2k** [comment]
> Is it possible something in your load balancer is stripping query params off the request?

LoadBalancer is the first place I would look.  I'm not a wss expert, but you could try mocking up a call to `oapi/v1/users/~` through the websocket to see which user the openshift api server thinks you are.

**Comment by deads2k** [comment]
Calling the `user/~` endpoint would let you know if the server is getting your token and you simply don't have rights. That would be unusual, but possible if something is missing in your roles.

**Comment by jwforres** [comment]
From the network trace in the initial screenshot I can see the token was
valid, the regular https requests were succeeding, only the wss requests
were failing.

**Comment by Mangirdaz** [comment]
Yes. The difference is that for HTTP you use in the Header:
```
Authorization:Bearer GalvMr6mknly-SOslIj0FjXg_6nqZLSf0ZcNF4kyofQ
```

and for WSS:
```
watch:true
access_token:GalvMr6mknly-SOslIj0FjXg_6nqZLSf0ZcNF4kyofQ
```

My JS skills are rusty but im trying to recreate this with simple:
```
....
var wsUri = "wss://console.int.xxx/oapi/v1/projects?watch=true";  
websocket = new WebSocket(wsUri,  ["access_token", "GalvMr6mknly-SOslIj0FjXg_6nqZLSf0ZcNF4kyofQ"]);
...
```
And getting undefined. will try dig into LB more and if found smth will drop here :/

**Comment by Mangirdaz** [comment]
Got involved with AVI. So apparently Openshift never responds with "upgraded" WS connection and instead gives 403.

Tried checking Logs on api by increasing log level to 4 (/etc/sysconfig/atomic-openshift-master-api) but cant see any request being logged.

Where does Logs for oapi/api on backed being logged?

**Comment by deads2k** [comment]
> Where does Logs for oapi/api on backed being logged?

Take a look here: https://docs.openshift.org/latest/install_config/master_node_configuration.html#master-node-config-audit-config .  Enabling the audit logging will show every connection.

**Comment by Mangirdaz** [comment]
failing setup:
AVI Loadbalancer is presenting certificates and doing SSL-ReEncryption. During SSL reencryption backend does not recognize token/cookie as valid and gives 403. This is why session is never upgraded to WS.

Fix:
We removed certificates from AVI and placed them on Openshift/Kubernetes itself and reconfigured AVI tp be L4/Passthrouth loadbalancer. 
We tried same thing with HAproxy re-encryption, and everything worked out of the box. 

Question:
Do you do any session pinning based on SSL certificate? or anything under these lines?  

Closing this one.

**Comment by Mangirdaz** [comment]
Fixed in AVI 16.3.4.
